art.hpp

```diff
@@ -124,7 +193,43 @@ class db final {
 
     // The stack is made up of tuples containing the node pointer, the
     // key_byte, and the child_index.
-    using stack_entry = detail::inode_base::iter_result;
+    //
+    // Note: The node is a pointer to either an internal node or a
+    // leaf.
+    //
+    // Note: The key_byte is the byte from the key that was consumed
+    // as you step down to the child node. This is the same as the
+    // child index (if you convert std::uint8_t to std::byte) for N48
+    // and N256, but it is different for N4 and N16 since they use a
+    // sparse encoding of the keys.  It is represented explicitly to
+    // avoid searching for the key byte in the N48 and N256 cases.
+    //
+    // Note: The child_index is the index of the child node within
+    // that internal node (except for N48, where it is the index into
+    // the child_indexes[] and is in fact the same data as the key
+    // byte).  Overflow for the child_index can only occur for N48 and
+    // N256.  When overflow happens, the iter_result is not defined
+    // and the outer std::optional will return false.
+    struct stack_entry : public inode_base::iter_result {
+      // The [prefix_len] is the number of bytes in the key prefix for
+      // [node].  When the node is pushed onto the stack, we also push
+      // [prefix_bytes] plus the [key_byte] onto a key_buffer.  We
+      // track how many bytes were pushed here (not including the
+      // key_byte) so we can pop off the correct number of bytes
+      // later.
+      //
+      // TODO(laurynas): push this field to a common parent? Shared with
```

Comment:
There is no such common parent class now.  We could introduce one, but it's not really worth it IMHO.

Reply:
@thompsonbry , I mean, shouldn't this field move to `inode_base::iter_result`?